File: simulate_matchups.mjs

import { GameState } from './js/game-engine.js';

// --- AI LOGIC ADAPTER (Mirror of game-engine.js for Player 1) ---
// We need to implement P1 logic because game-engine only has P2 logic built-in.
// P1 logic is identical but sees 'score1' as 'my score' and 'score2' as 'enemy score'.

function getP1Move(game, difficulty) {
    const lanes = ['left', 'center', 'right'];

    // RANDOM
    if (difficulty === 0) {
        return lanes[Math.floor(Math.random() * lanes.length)];
    }

    // EASY
    if (difficulty === 1) {
        // Avoid full lanes (>=2 cards) or diff < -30
        const okLanes = lanes.filter(l => {
            const laneData = game.lanes[l];
            // Perspective Flip: P1 wants score1 > score2. 
            // Diff for P1 is (score1 - score2). 
            // The original code uses (score2 - score1) for P2.
            const p1Diff = laneData.score1 - laneData.score2;
            const p1Cards = laneData.history.filter(h => h.player === 'p1').length;
            return p1Diff > -30 && p1Cards < 2;
        });
        if (okLanes.length > 0) return okLanes[Math.floor(Math.random() * okLanes.length)];

        const withSpace = lanes.filter(l => {
            const p1Cards = game.lanes[l].history.filter(h => h.player === 'p1').length;
            return p1Cards < 2;
        });
        if (withSpace.length > 0) return withSpace[Math.floor(Math.random() * withSpace.length)];
        return lanes[Math.floor(Math.random() * lanes.length)];
    }

    // ELITE (PRO / HARD+)
    const canPeek = (difficulty === 3);

    // 1. Evaluate Lane States (Perspective Flip)
    const evaluations = lanes.map(lane => {
        const l = game.lanes[lane];
        const diff = l.score1 - l.score2; // P1 Advantage
        const p1Cards = l.history.filter(h => h.player === 'p1').length;
        const p2Cards = l.history.filter(h => h.player === 'p2').length;

        // Status
        const isLost = diff < -25;
        const isSecure = diff > 25;

        return { lane, diff, p1Cards, p2Cards, isLost, isSecure };
    });

    // 2. Filter out "Lost" lanes
    let playable = evaluations.filter(e => !e.isLost);
    if (playable.length === 0) {
        evaluations.sort((a, b) => b.diff - a.diff);
        return evaluations[0].lane;
    }

    // 3. Sort by Advantage (Best first)
    playable.sort((a, b) => b.diff - a.diff);

    // 4. Select Target Strategy
    const targetLanes = playable.slice(0, 2);

    // 5. PEEKING LOGIC
    let isHighCard = false;
    if (canPeek) {
        // P1 Deck Peek
        const myCardValue = game.player1Deck.length > 0 ? game.player1Deck[0].value : 5;
        isHighCard = myCardValue >= 7;
    }

    if (targetLanes.length === 0) return 'center';

    if (targetLanes.length > 1) {
        const best = targetLanes[0];
        const second = targetLanes[1];

        // Priority: If secondary is exposed (0 cards), grab it!
        // (Checking 'p2Cards' which is the OPPONENT count in that lane for us?)
        // In original code: `if (second.p1Cards === 0)`. 'p1Cards' was 'my cards'.
        // So here we check 'p1Cards' (our cards).
        if (second.p1Cards === 0) return second.lane;

        // Priority: If best is secure, focus on second
        if (best.isSecure) return second.lane;

        // Priority: Peek High Card
        if (canPeek && isHighCard) return second.lane;

        // Priority: Balance
        return second.lane;
    } else {
        return targetLanes[0].lane;
    }
}

// --- SIMULATION RUNNER ---

function simulateMatch(p1Diff, p2Diff, iterations) {
    let p1Wins = 0;
    let p2Wins = 0;
    let draws = 0;

    for (let i = 0; i < iterations; i++) {
        const game = new GameState();
        game.setDifficulty(p2Diff); // Set Engine's P2 Logic

        while (!game.isGameOver) {
            // P1 Decision
            const p1Lane = getP1Move(game, p1Diff);

            // P2 Decision is handled INSIDE playRound automatically based on setDifficulty
            // BUT game-engine.js `playRound` decides cpuLane internally.
            // We just call `playRound(p1Lane)`.
            game.playRound(p1Lane);
        }

        // Tally
        let p1Score = 0;
        let p2Score = 0;
        ['left', 'center', 'right'].forEach(l => {
            if (game.lanes[l].score1 > game.lanes[l].score2) p1Score++;
            else if (game.lanes[l].score2 > game.lanes[l].score1) p2Score++;
        });

        if (p1Score > p2Score) p1Wins++;
        else if (p2Score > p1Score) p2Wins++;
        else draws++;
    }

    return { p1Wins, p2Wins, draws };
}

// --- EXECUTION ---

const LEVELS = ['RANDOM', 'EASY', 'PRO', 'HARD+'];
const ITERATIONS = 1000;

console.log(`\n🤖 WAR.0 AI MATCHUP SIMULATION (${ITERATIONS} games per matchup)\n`);
console.log('P1 (Rows) vs P2 (Cols) Win Rates (for P1)');
console.log('------------------------------------------------');

// Header
process.stdout.write('P1 \\ P2'.padEnd(12));
LEVELS.forEach(l => process.stdout.write(l.padEnd(10)));
console.log('\n' + '-'.repeat(60));

// Rows
for (let p1 = 0; p1 < 4; p1++) {
    process.stdout.write(LEVELS[p1].padEnd(12));

    for (let p2 = 0; p2 < 4; p2++) {
        const result = simulateMatch(p1, p2, ITERATIONS);
        const winRate = ((result.p1Wins / ITERATIONS) * 100).toFixed(1) + '%';
        process.stdout.write(winRate.padEnd(10));
    }
    console.log('');
}
console.log('------------------------------------------------\n');

File: simulate_matchups.py

import random
import statistics

# --- CONSTANTS ---
SUITS = ['♠', '♥', '♣', '♦']
RANKS = ['2', '3', '4', '5', '6', '7', '8', '9', '10']

class Card:
    def __init__(self, suit, rank, value):
        self.suit = suit
        self.rank = rank
        self.value = value

class GameState:
    def __init__(self):
        self.deck1 = self.create_deck()
        self.deck2 = self.create_deck()
        self.lanes = {
            'left': {'score1': 0, 'score2': 0, 'p1_cards': 0, 'p2_cards': 0, 'history': []},
            'center': {'score1': 0, 'score2': 0, 'p1_cards': 0, 'p2_cards': 0, 'history': []},
            'right': {'score1': 0, 'score2': 0, 'p1_cards': 0, 'p2_cards': 0, 'history': []}
        }
        self.is_game_over = False

    def create_deck(self):
        cards = []
        for suit in SUITS:
            for i, rank in enumerate(RANKS):
                cards.append(Card(suit, rank, i + 2))
        random.shuffle(cards)
        return cards

    def play_round(self, p1_lane, p2_lane):
        if not self.deck1 or not self.deck2:
            self.is_game_over = True
            return

        c1 = self.deck1.pop(0)
        c2 = self.deck2.pop(0)

        l1 = self.lanes[p1_lane]
        l1['score1'] += c1.value
        l1['p1_cards'] += 1
        l1['history'].append({'player': 'p1', 'val': c1.value})

        l2 = self.lanes[p2_lane]
        l2['score2'] += c2.value
        l2['p2_cards'] += 1
        l2['history'].append({'player': 'p2', 'val': c2.value})

        if not self.deck1:
            self.is_game_over = True

# --- AI LOGIC (Matches JS Logic) ---

def get_ai_move(game, difficulty, is_p1):
    lanes_keys = ['left', 'center', 'right']

    # RANDOM (0)
    if difficulty == 0:
        return random.choice(lanes_keys)

    # EASY (1)
    if difficulty == 1:
        candidates = []
        for l in lanes_keys:
            lane = game.lanes[l]
            diff = (lane['score1'] - lane['score2']) if is_p1 else (lane['score2'] - lane['score1'])
            my_cards = lane['p1_cards'] if is_p1 else lane['p2_cards']
            if diff > -30 and my_cards < 2:
                candidates.append(l)
        if candidates: return random.choice(candidates)
        space_fallback = [l for l in lanes_keys if (game.lanes[l]['p1_cards' if is_p1 else 'p2_cards'] < 2)]
        if space_fallback: return random.choice(space_fallback)
        return random.choice(lanes_keys)

    # PRO (2) / HARD+ (3)
    can_peek = (difficulty == 3)
    
    # 1. Evaluate
    evals = []
    for l in lanes_keys:
        lane = game.lanes[l]
        diff = (lane['score1'] - lane['score2']) if is_p1 else (lane['score2'] - lane['score1'])
        my_cards = lane['p1_cards'] if is_p1 else lane['p2_cards']
        enemy_cards = lane['p2_cards'] if is_p1 else lane['p1_cards']

        is_lost = diff < -25
        is_secure = diff > 25
        
        evals.append({
            'lane': l,
            'diff': diff,
            'my_cards': my_cards,
            'enemy_cards': enemy_cards,
            'is_lost': is_lost,
            'is_secure': is_secure
        })

    # 2. Filter Lost
    playable = [e for e in evals if not e['is_lost']]
    if not playable:
        evals.sort(key=lambda x: x['diff'], reverse=True)
        return evals[0]['lane']

    # 3. Sort by Advantage (Best First)
    playable.sort(key=lambda x: x['diff'], reverse=True)

    # 4. Target Strategy
    targets = playable[:2]
    
    # 5. Peeking
    my_card_value = 6
    if can_peek:
        my_deck = game.deck1 if is_p1 else game.deck2
        if my_deck: my_card_value = my_deck[0].value
        else: my_card_value = 5

    if not targets: return 'center'

    if len(targets) > 1:
        best = targets[0]
        second = targets[1]

        # Priority: If secondary is exposed (0 cards), grab it!
        if second['enemy_cards'] == 0: return second['lane']
        
        # Priority: If best is already secure, focus everything on second
        if best['is_secure']: return second['lane']
        
        # HARD+ INTELLIGENCE (EFFICIENCY LOGIC)
        if can_peek:
            # 1. RESCUE THE LEADER
            if best['diff'] < 0:
                 return best['lane']
            
            # 2. SWEEP STRATEGY
            if second['diff'] > 0 and len(playable) > 2:
                 third = playable[2]
                 if third['diff'] + my_card_value > 0:
                     return third['lane']
            
            # 3. HIGH CARD SNIPE
            if my_card_value >= 8:
                 return second['lane']
        
        # PRO STRATEGY / DEFAULT
        return second['lane']
    
    else:
        return targets[0]['lane']

# --- SIMULATION LOOP ---

def simulate(p1_diff, p2_diff, iterations=1000):
    p1_wins = 0
    p2_wins = 0
    draws = 0

    for _ in range(iterations):
        game = GameState()
        while not game.is_game_over:
            m1 = get_ai_move(game, p1_diff, True)
            m2 = get_ai_move(game, p2_diff, False)
            game.play_round(m1, m2)

        # Result
        s1 = 0
        s2 = 0
        for l in game.lanes.values():
            if l['score1'] > l['score2']: s1 += 1
            elif l['score2'] > l['score1']: s2 += 1
        
        if s1 > s2: p1_wins += 1
        elif s2 > s1: p2_wins += 1
        else: draws += 1

    return p1_wins / iterations

# --- MAIN ---

LEVELS = ['RANDOM', 'EASY', 'PRO', 'HARD+']
print(f"\n🤖 WAR.0 AI MATCHUP SIMULATION (1000 games per matchup)\n")
header_start = "P1 \\ P2"
print(f"{header_start:<12} {'RANDOM':<10} {'EASY':<10} {'PRO':<10} {'HARD+':<10}")
print("-" * 60)

for p1_i, p1_name in enumerate(LEVELS):
    row_str = f"{p1_name:<12}"
    for p2_i, p2_name in enumerate(LEVELS):
        rate = simulate(p1_i, p2_i)
        row_str += f"{rate*100:.1f}%".ljust(11)
    print(row_str)

print("-" * 60 + "\n")
